src/main/java/io/strimzi/kproxy/internal/KafkaProxyInitializer.java

```diff
@@ -78,4 +80,28 @@ public void initChannel(SocketChannel ch) {
         }
         ch.pipeline().addLast(new KafkaProxyFrontendHandler(remoteHost, remotePort, correlation, interceptorProvider, logNetwork, logFrames));
     }
+
+    private static class RequestHandlerAdapter extends ChannelInboundHandlerAdapter {
+
+        private final Interceptor interceptor;
+
+        public RequestHandlerAdapter(Interceptor requestInterceptor) {
+            this.interceptor = requestInterceptor;
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+            if (msg instanceof DecodedRequestFrame) {
+                DecodedRequestFrame decodedFrame = (DecodedRequestFrame) msg;
+                if (interceptor.shouldDecodeRequest(decodedFrame.apiKey(), decodedFrame.apiVersion())) {
+                    interceptor.requestHandler().handleRequest(decodedFrame, new DefaultHandlerContext(ctx));
+                }
+            }
+
+            // TODO: when to release ByteBuf allocated via DefaultHandlerContext?
```

Comment:
@franz1981 That's the part I'm unsure about. Can I call `release()` here for all the `ByteBuf` that were allocated through `DefaultHandlerContext` (which means I need to keep track of them there)? The buffer still is referenced by the `msg` object which is passed on to the next handler, but there are no references by the Kafka builder classes at this point.

Reply:
Whatever flow through the pipeline (in the form of the `msg` passing across channel handlers) is already released by Netty at "the end", see https://netty.io/wiki/reference-counted-objects.html#reference-counting-in-channelhandler for more info